File: Aulas/Modulo1/Pilhas_Filas/150. Evaluate reverse polish notation.cs
//https://leetcode.com/problems/evaluate-reverse-polish-notation/

/*
Estratégia: percorrer o array de tokens empilhando os operandos.
Quando encontrar um operador, retira os dois últimos operandos da pilha e realiza a operação, empilhando o resultado.
Como a entrada é garantida ser uma expressão válida, ao final das operações a pilha conterá um único valor, que é a avaliação da expressão de entrada.
*/

public class Solution {
    public int EvalRPN(string[] tokens) {
        
        Stack<int> pilha = new Stack<int>();

        foreach (string token in tokens) 
        {
            if (!"+-*/".Contains(token)) 
            {
                pilha.Push(Int32.Parse(token));
                continue;
            }

            int var1 = pilha.Pop();
            int var2 = pilha.Pop();
            char op = Char.Parse(token);

            pilha.Push(CalculaOperacao(var1, var2, op));
        }

        return pilha.Pop();
    }

    private int CalculaOperacao(int valor1, int valor2, char operador){
        if (operador == '+'){
            return valor2 + valor1;
        } else if (operador == '-'){
            return valor2 - valor1;
        } else if (operador == '*'){
            return valor2 * valor1;
        } else if (operador == '/'){
            return valor2/valor1;
        } else{
            return 0;
        }

    }
}

File: Aulas/Modulo1/Pilhas_Filas/150. Evaluate Reverse Polish Notation.py
#https://leetcode.com/problems/evaluate-reverse-polish-notation/

#ATENCAO para usar / e não //, pois o // arredonda para baixo

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        operands = ['+','-','*','/']
        stack = []

        for i in tokens:
            if i not in operands:
                stack.append(int(i))
            else:
                primeiro = stack.pop(-1)
                segundo = stack.pop(-1)
                #print(f'{segundo} {str(i)} {primeiro}')
                if i == '+':
                    stack.append(segundo + primeiro)
                elif i == '-':
                    stack.append(segundo - primeiro)
                elif i == '*':
                    stack.append(segundo * primeiro)
                elif i == '/':
                    stack.append(int(segundo / primeiro)) 

        return stack.pop(-1)